sha2/build.rs

```diff
@@ -1,14 +1,34 @@
+use std::collections::HashSet;
+
 fn main() {
     use std::env;
 
     let target_arch = env::var("CARGO_CFG_TARGET_ARCH").unwrap_or_default();
     let target_vendor = env::var("CARGO_CFG_TARGET_VENDOR").unwrap_or_default();
+    let features: HashSet<String> = env::var("CARGO_CFG_TARGET_FEATURE")
+        .unwrap_or_default()
+        .split(',')
+        .map(|x| x.to_string())
+        .collect();
 
     let mut build256 = cc::Build::new();
     let (sha256_path, sha512_path) = if target_arch == "x86" {
         ("src/sha256_x86.S", "src/sha512_x86.S")
     } else if target_arch == "x86_64" {
-        ("src/sha256_x64.S", "src/sha512_x64.S")
+        let sha512 = if features.contains("avx2") {
+            "src/sha512_x64_avx2.S"
+        } else {
+            "src/sha512_x64.S"
+        };
+        // Prioritizing aes-ni, cause it's fastest
+        let sha256 = if features.contains("aes") {
+            "src/sha256_x64_ni.S"
+        } else if features.contains("avx2") {
+            "src/sha256_x64_avx2.S"
+        } else {
+            "src/sha256_x64.S"
+        };
```

Comment:
I'm guessing this is SHA-NI, which is a completely different feature set from AES-NI.

There is no AES involved in SHA.

Reply:
And regarding SHA-NI, we already support it: https://github.com/RustCrypto/hashes/pull/224

Reply:
My bad

Reply:
> And regarding SHA-NI, we already support it: [RustCrypto/hashes#224](https://github.com/RustCrypto/hashes/pull/224)

It seems like it's only for aarch64, no?

Reply:
SHA-NI is an x86 feature and that PR is for x86, AFAICT

Reply:
The linked PR simply has made priority of the intrinsics-based implementation higher than the linked PR.  The implementation [itself](https://github.com/RustCrypto/hashes/blob/master/sha2/src/sha256/x86.rs) was added in RustCrypto/hashes#167. BTW we probably should use a better name, something like `x86_shani.rs`. Also are you sure it's correct to call it SHA-NI? I don't remember it being referenced like this, usually it's simply "SHA extension".

Assuming performance of the intrinsics-based implementation is similar to the assembly in this PR, I think it's not worth to keep both and we should prefer the intrinsics-based implementation.

Reply:
I've checked, intrinsic based version takes exactly the same time. So it seems, that porting sha512 and sha256 to avx2 intrinsic will give much more profit

Reply:
> Also are you sure it's correct to call it SHA-NI? I don't remember it being referenced like this, usually it's simply "SHA extension".

SHA-NI is a common abbreviation for it and used by Intel as well, for example:

https://software.intel.com/content/www/us/en/develop/documentation/ipp-crypto-reference/top/one-way-hash-primitives.html

> These structures check for support of the SHA-NI instruction set at run time and choose the implementation of an algorithm depending on the outcome of the check.